// paprika-api/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var rp = require("request-promise-native");
var PaprikaApi = /** @class */ (function () {
    function PaprikaApi(email, password) {
        this.baseUrl = "https://www.paprikaapp.com/api/v1/sync/";
        this.email = email;
        this.password = password;
    }
    PaprikaApi.prototype.resource = function (endpoint) {
        var options = {
            auth: {
                user: this.email,
                pass: this.password
            },
            method: 'GET',
            baseUrl: this.baseUrl,
            uri: endpoint,
            json: true,
            headers: {
                'User-Agent': 'PaprikaApi NodeJS library'
            },
            transform: function (body) {
                return body.result;
            }
        };
        return rp(options);
    };
    PaprikaApi.prototype.bookmarks = function () {
        return this.resource('bookmarks');
    };
    PaprikaApi.prototype.categories = function () {
        return this.resource('categories');
    };
    PaprikaApi.prototype.groceries = function () {
        return this.resource('groceries');
    };
    PaprikaApi.prototype.meals = function () {
        return this.resource('meals');
    };
    PaprikaApi.prototype.menus = function () {
        return this.resource('menus');
    };
    PaprikaApi.prototype.menuItems = function () {
        return this.resource('menuitems');
    };
    PaprikaApi.prototype.pantry = function () {
        return this.resource('pantry');
    };
    PaprikaApi.prototype.recipes = function () {
        return this.resource('recipes');
    };
    PaprikaApi.prototype.recipe = function (recipeUid) {
        return this.resource('recipe/' + recipeUid);
    };
    PaprikaApi.prototype.status = function () {
        return this.resource('status');
    };
    return PaprikaApi;
}());
exports.PaprikaApi = PaprikaApi;
